c2rust-analyze/src/c_void_casts.rs

```diff
@@ -274,16 +287,88 @@ impl<'tcx> CVoidCasts<'tcx> {
-        };
-        self.to.contains(lhs)
-            || self.from.contains(match get_cast_place(rv) {
-                None => return false,
-                Some(it) => it,
-            })
+    pub fn should_skip_stmt(&self, loc: &Location) -> bool {
+        self.to.contains(loc) || self.from.contains(loc)
+    }
+
+    fn is_place_local_modified(p: &Place, stmt: &Statement) -> bool {
+        if !p.projection.is_empty() {
+            return false;
+        }
+
+        let local = p.local;
+
+        if let StatementKind::Assign(asgn) = stmt.kind.clone() {
+            let (lhs, rv) = *asgn;
+            if lhs.local == local {
+                return true;
+            }
+
+            use Rvalue::*;
+            let rv_place = match rv {
+                Use(op) => op.place(),
+                Repeat(op, _) => op.place(),
+                Ref(_, _, p) => Some(p),
+                // ThreadLocalRef
+                AddressOf(_, p) => Some(p),
+                Len(p) => Some(p),
+                Cast(_, op, _) => op.place(),
+                // BinaryOp
+                // CheckedBinaryOp
+                // NullaryOp
+                UnaryOp(_, op) => op.place(),
+                Discriminant(p) => Some(p),
+                // Aggregate
+                ShallowInitBox(op, _) => op.place(),
+                _ => None,
+            };
+
+            if let Some(rv_local) = rv_place.map(|p| p.local) {
+                return local == rv_local;
+            }
+        }
+
+        false
+    }
+
+    fn find_first_cast_succ_block<F: FnMut(&Statement<'tcx>) -> Option<CVoidCast<'tcx>>>(
+        mut get_cast: F,
+        statements: &[Statement<'tcx>],
+    ) -> Option<(usize, CVoidCast<'tcx>)> {
+        // For [`CVoidCastDirection::From`], we only count
+        // a cast from `*c_void` to an arbitrary type in the subsequent block,
+        // searching forward.
+        for (sidx, stmt) in statements.iter().enumerate() {
+            if let StatementKind::StorageDead(_) = stmt.kind {
+                continue;
+            }
+            if let Some(cast) = get_cast(stmt) {
+                return Some((sidx, cast));
+            } else {
+                break;
+            }
+        }
+
+        None
+    }
+
+    fn find_last_cast_curr_block<F: FnMut(&Statement<'tcx>) -> Option<CVoidCast<'tcx>>>(
+        mut get_cast: F,
+        statements: &[Statement<'tcx>],
+        place: &Place<'tcx>,
+    ) -> Option<(usize, CVoidCast<'tcx>)> {
+        // For [`CVoidCastDirection::From`], we only count
+        // a cast to `*c_void` from an arbitrary type in the same block,
+        // searching backwards.
+        for (sidx, stmt) in statements.iter().enumerate().rev() {
+            let cast = get_cast(stmt);
+            if let Some(cast) = cast {
+                return Some((sidx, cast));
+            } else if Self::is_place_local_modified(place, stmt) {
+                return None;
+            }
+        }
+        None
```

Comment:
Can you explain (in docs, too) what this is doing?

Reply:
ed3c93a9